### ISSUE TITLE
no-unused-vars reported for a used var

### ISSUE THREAD

**Issue by reintroducing** [eslint/eslint#3129]
First off let me start by saying that ESLint is a *fantastic* tool and while incorporating it into the SpotHero website I've been amazed at how many times it has caught something that I did not think was possible.  You have done a fantastic job in developing this and the team should be commended for it.

That being said, here are some details about the issue at hand.  I have a Browserify module that defines some properties at the start, like this (don't judge us for mixing requires with props and function declarations :):

```
'use strict';

var _ = require('underscore'),
    moment = require('moment-timezone'),
    AbstractView = require('abstract-view'),

    DATETIME_FORMAT = 'MM/DD/YYYY',

    ...
    startDatePicker,
    ...
```

I then export a module that extends AbstractView which is an extension BackboneView itself.  Inside the render method I have a _.delay call which defines the following in its callback:

```
startDatePicker = $startDate.pickadate();
```

ESLint reports that startDatePicker is never used:

```
20:4  error  startDatePicker is defined but never used  no-unused-vars
```

I've had this happen in a few files now and added the ESLint ignore line comment targeting that rule but it's getting to a point where I'm curious if it has something to do with the way I'm defining my modules or if this is really an ESLint issue that I want to point out to you guys in case it needs to be fixed.

Please advise if this is a "me" problem and what could be the cause of it.  If you'd like I'd be more than happy to send the whole file over to you or let me know if there is any way I can help in tracking this down otherwise.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by IanVS** [comment]
Maybe I'm missing it, but how is `startDatePicker` being used?  I only see where you've defined it.

**Comment by reintroducing** [comment]
version: gulp-eslint 0.15.0 which uses 0.24.0
config: 
```
"no-unused-vars": [2, {
    "vars": "all",
    "args": "none"
}]
```

**Comment by reintroducing** [comment]
@IanVS The second code block shows how it is being used. It gets assigned to a pickadate().

**Comment by reintroducing** [comment]
@IanVS I think I just realized what you're asking.  Just assigning it does not mark it as used, actually then using it does, which makes complete sense because I just realized it does not get used after it gets assigned.  Is that what you're leading at?

**Comment by IanVS** [comment]
It is being initialized there, but not actually *used*.  It's like saying:

```
var a = 2;
```

If you're not going to do something with `a`, why create it as a variable?  So, it seems to me like the rule is behaving correctly and flagging what it is supposed to.

**Comment by reintroducing** [comment]
@IanVS Yep that makes absolute sense now.  I did not realize how damn smart this thing was, it's catching me not even using my assigned variables.  That's damn impressive!

**Comment by IanVS** [comment]
Glad you like it, and thanks for taking the time to submit an issue.

**Comment by reintroducing** [comment]
@IanVS So I'm going through my code just now to get rid of the properties which this applied to which I had marked for ignoring.  There is one scenario I'm coming across that would be slightly misleading.

```
...
mapView, // eslint-disable-line no-unused-vars
...
mapView = new MapView(mapViewOptions);
```

MapView is a Backbone view that is self rendering (meaning it does not need render() called on it after it is initialized) so it does need instantiation but no other calls to it are made.  I've always created a variable for it in this way "just in case" but also because if I'm not mistaken ESLint has another rule which would error if it was just written as `new MapView()` without being assigned to anything (which is a preference of mine as well so I keep that rule on).

Is there a way to set this up so that it does not error when a constructor is used versus just a plain assignment? Or is this just a case of where I need to disable the line and be done with it?

**Comment by IanVS** [comment]
I'm no expert in Backbone (or anything else for that matter), but I believe that ESLint is still doing its job in this case.  From the [no-new docs](http://eslint.org/docs/rules/no-new):

> This rule is aimed at maintaining consistency and convention by disallowing constructor calls using the new keyword that do not assign the resulting object to a variable.

So, it seems that the way `MapView` is used in Backbone does not conform to this particular convention, and I would recommend disabling `no-new` for that line, or for your project in general if this is a common Backbone pattern.

**Comment by reintroducing** [comment]
@IanVS So that would make sense except I have that rule turned on and in this case I am indeed assigning the resulting object to a variable.  Granted I am not using it later but I do need to instantiate it and I'm assigning it because I believe in that being a good convention.  Maybe I'm missing something here (entirely and likely possible)?

**Comment by IanVS** [comment]
Yes, so:
 you should not be getting `no-new` errors with `mapView = new MapView(mapViewOptions);`

with this config:
```
{
  rules: {
    no-new: 2,
    no-unused-vars: 2
  }
}
```

You will get:
```js
var mapView = new MapView(mapViewOptions);
// -> no-unused-vars error

new MapView(mapViewOptions);
// -> no-new error
```

I am suggesting to change your config to:
```
{
  rules: {
    no-new: 0,
    no-unused-vars: 2
  }
}
```

And your code to:
```js
new MapView(mapViewOptions);
```

That is one way you could avoid linting errors, by disabling the rule which does not fit with your coding style.

**Comment by reintroducing** [comment]
Gotcha. I'd rather not disable that rule (no-new) because I do want to enforce it since my preference is not to write code that way so I think the best solution for my case is just the disable line I originally had for this particular case.  

Thanks again for your help and keep up the great work.
